File: Assembler.hs
module Assembler (assemble) where

import Ast (Instruction(..))
import Data.List (nub, mapAccumR)
import Data.Maybe (catMaybes, fromJust)
import Data.Bits (shiftR)

assemble :: [Instruction] -> String
assemble ins = let codesTable = numberLabels $ getLabels ins
               in  concat $ map (translate codesTable) ins

s = ' '
t = '\t'
f = '\n'

translate :: [(String, Integer)] -> Instruction -> String

translate _ (Push n)        = s : s : (toWhitespace n)
translate _ (Dup)           = s : f : s : []
translate _ (Copy n)        = s : t : s : (toWhitespace n)
translate _ (Swap)          = s : f : t : []
translate _ (Pop)           = s : f : f : []
translate _ (Slide n)       = s : t : f : (toWhitespace n)

translate _ (Add)           = t : s : s : s : []
translate _ (Sub)           = t : s : s : t : []
translate _ (Mul)           = t : s : s : f : []
translate _ (Div)           = t : s : t : s : []
translate _ (Mod)           = t : s : t : t : []

translate _ (Store)         = t : t : s : []
translate _ (Load)          = t : t : t : []

translate codes (Label lbl) = f : s : s : (translateLabel codes lbl)
translate codes (Call lbl)  = f : s : t : (translateLabel codes lbl)
translate codes (Jump lbl)  = f : s : f : (translateLabel codes lbl)
translate codes (Jz lbl)    = f : t : s : (translateLabel codes lbl)
translate codes (Jn lbl)    = f : t : t : (translateLabel codes lbl)
translate _     (Ret)       = f : t : f : []
translate _     (End)       = f : f : f : []

translate _ (Pc)            = t : f : s : s : []
translate _ (Pn)            = t : f : s : t : []
translate _ (Rc)            = t : f : t : s : []
translate _ (Rn)            = t : f : t : t : []

translateLabel :: [(String, Integer)] -> String -> String
translateLabel labelCodes label = toWhitespace $ fromJust $ lookup label labelCodes

toWhitespace :: Integer -> String
toWhitespace i = if i >= 0
                    then (map bitCode $ 0:(bits i)) ++ "\n"
                    else (map bitCode $ 1:(bits (-i))) ++ "\n"
               where bits 0 = [0]
                     bits n | odd n     = (bits $ shiftR n 1) ++ [1] 
                            | otherwise = (bits $ shiftR n 1) ++ [0]
                     bitCode 0 = ' '
                     bitCode 1 = '\t'
                     
numberLabels :: [String] -> [(String, Integer)]
numberLabels lbs = snd $ mapAccumR enumerate 0 lbs
                 where enumerate i item = (i+1, (item, i))    

getLabels :: [Instruction] -> [String]
getLabels ins = nub $ catMaybes $ map getLabel ins

getLabel :: Instruction -> Maybe String
getLabel (Label lb) = Just lb
getLabel (Call lb) = Just lb
getLabel (Jump lb) = Just lb
getLabel (Jz lb) = Just lb
getLabel (Jn lb) = Just lb
getLabel _ = Nothing

File: assembler.py

class AssemblerError(Exception):
    pass

s = ' '
t = '\t'
n = '\n'

stack = s
arith = t + s
heap  = t + t
flow  = n
io    = t + n

OPCODES = dict(
    push  = stack + s,
    dup   = stack + n + s,
    copy  = stack + t + s,
    swap  = stack + n + t,
    pop   = stack + n + n,
    slide = stack + t + n,
    
    add = arith + s + s,
    sub = arith + s + t,
    mul = arith + s + n,
    div = arith + t + s,
    mod = arith + t + t,
    
    store = heap + s,
    load  = heap + t,
    
    label = flow + s + s,
    call  = flow + s + t,
    jump  = flow + s + n,
    jz    = flow + t + s,
    jn    = flow + t + t,
    ret   = flow + t + n,
    end   = flow + n + n,
    
    pc = io + s + s,
    pn = io + s + t,
    rc = io + t + s,
    rn = io + t + t,
)    

def numberToWhitespace(value):
    
    if value < 0:
        sign = 1
    else:
        sign = 0
        
    value = abs(value)
    
    if value == 0:
        revbits = [0]
    else:
        revbits = []
        while value != 0:
            bit = value & 0x01
            revbits.append(bit)
            value = value >> 1
    
    revbits.append(sign)
    
    bits = reversed(revbits)
    binstr = ''.join(map(str, bits))
    
    wsstr = binstr.replace('0', s).replace('1', t) + '\n'
        
    return wsstr
    
def translateToWhitespace(assembly):
    
    assembledOpcodes = []
    labelsDefined = []
    
    for (command, operand) in assembly:
        if command == "label":
            if operand in labelsDefined:
                raise AssemblerError("Multiple definitions of label " + operand + " found.")
            else:
                labelsDefined.append(operand)
                
    for (command, operand) in assembly:
        
        opcode = OPCODES[command]
        
        if command in ["push", "copy", "slide"]:
            opcode += numberToWhitespace(operand)
        elif command in ["label", "call", "jump", "jz", "jn"]:
            if operand in labelsDefined:
                opcode += numberToWhitespace(labelsDefined.index(operand))
            else:
                raise AssemblerError("Jump to undefined label " + operand + " found.")
            
        assembledOpcodes.append(opcode)
    
    whitespace = ''.join(assembledOpcodes)
    return whitespace
    
